examples/larger-examples/datalog/ford-fulkerson.flix

```diff
@@ -0,0 +1,649 @@
+///
+/// The Ford-Fulkerson algorithm finds the maximum flow of a flow network.
+/// Here it is implemented using a combination of functional programming
+/// and datalog.
+///
+pub def main(): Unit \ IO =
+    let g1 = FordFulkerson.exampleGraph01() |> FordFulkerson.maxFlow(src = 0, dst = 5);
+    let g2 = FordFulkerson.exampleGraph02() |> FordFulkerson.maxFlow(src = 0, dst = 5);
+    let g3 = FordFulkerson.exampleGraph03() |> FordFulkerson.maxFlow(src = 0, dst = 3);
+    let g4 = FordFulkerson.exampleGraph04() |> FordFulkerson.maxFlow(src = 0, dst = 10);
+    println("Expected maximum flow 19, got ${g1}");
+    println("Expected maximum flow 20, got ${g2}");
+    println("Expected maximum flow 2000, got ${g3}");
+    println("Expected maximum flow 20, got ${g4}")
+
+mod FordFulkerson {
+
+    use Path.{Path, Bot};
+
+    ///
+    /// Returns the maximum flow from `src` to `dst` in the flow network `g`.
+    /// N.B.: `g` is a directed graph with upper bounds / capacity on the edges.
+    /// No pre-assigned flow is allowed.
+    ///
+    /// The following assumptions also apply:
+    /// - `src` and `dst` is connected in `g`
+    /// - `g` contains no negative cycles
+    /// - `g` is labeled in ascending order from `src` to `sink`
+    /// - The label of `src` has the lowest value in the graph
+    /// - The label of `dst` has the highest value in the graph
+    ///
+    pub def maxFlow(src: { src = a }, dst: { dst = a }, g: m[(a, a, UpperCapacity)]): Flow with Foldable[m], Order[a] =
+        def fordFulkerson(flowNetwork) = match augmentingPath(src, dst, flowNetwork) {
+            case None       => getMaxFlow(dst.dst, flowNetwork)
+            case Some(path) =>
+                let incr = minCapacity(path, flowNetwork);
+                let updatedNetwork = increaseFlow(path, incr, flowNetwork);
+                fordFulkerson(updatedNetwork)
+        };
+        // Init with 0 flow
+        fordFulkerson(zeroFlow(g))
+
+    ///
+    /// Returns a flow network with zero flow.
+    ///
+    def zeroFlow(g: m[(a, a, UpperCapacity)]): Vector[FlowNetwork[a]] with Foldable[m], Order[a] =
+        Foldable.toVector(g) |> Vector.map(match (x, y, z) -> (x, y, z, 0))
+
+    ///
+    /// Returns the sum of the flows directly connected to `dst`.
+    ///
+    def getMaxFlow(dst: a, g: m[FlowNetwork[a]]): Flow with Foldable[m], Order[a] =
```

Comment:
It is not exactly clear to me what this computes. I think its the sum of the flow on all ingoing edges to dst?

Reply:
Yes